src/wpas/wpa_controller.c

```diff
@@ -820,8 +820,9 @@ __wpa_controller_send_commandf(struct wpa_controller *ctrl, const char *name, ch
 int
 wpa_controller_qrcode(struct wpa_controller *ctrl, const char *dpp_uri, uint32_t *bootstrap_id)
 {
-    char reply[WPA_MAX_MSG_SIZE];
-    size_t reply_length = sizeof reply;
+    char reply[WPA_MAX_MSG_SIZE + 1];
+    explicit_bzero(reply, sizeof reply);
```

Comment:
Looking at the implementation of [wpa_ctrl_request()](https://w1.fi/cgit/hostap/tree/src/common/wpa_ctrl.c#n481), it's curious that in the usual case, the `reply` buffer is _not_ NULL terminated. However, in the case where a callback is provided for messages received while waiting for the original reply, that buffer _is_ NULL-terminated ([here](https://w1.fi/cgit/hostap/tree/src/common/wpa_ctrl.c#n567)).

Either way, we can tighten this up a little and avoid zero'ing the input buffers. The onus shouldn't be on each individual `wpa_controller_<command>` function to specify a NULL-terminated buffer, and in some cases, may not even be possible. Instead, we should terminate the reply buffer that was specified directly in `__wpa_controller_send_commandf` (eg. `reply[reply_length] = '\0'`). In that case, you can still bump the buffer size by 1 in the instances that call `__wpa_controller_send_commandf` to ensure they can receive the maximum reply size. Though if you want to keep the PR more lean, you can omit those too since they're not actually needed to fix this bug. Up to you.

Reply:
Ok I think I incorporated your suggestions!

Reply:
Left some comments on the changes.